File: 238-product-of-array-except-self/238-product-of-array-except-self.java

class Solution {
    public int[] productExceptSelf(int[] nums) {
        //nums =    [1,   2, 3, 4]
        //prefix =  [1,   1, 2, 6]
        //postfix = [24, 12, 4, 1]
        //result =  [24, 12, 8, 6]
        
        int[] res = new int[nums.length];
        res[0] = 1;
        
        //prefix
        for(int i = 1; i < res.length; i++){
            res[i] = nums[i-1] * res[i-1];
        }
        
        //prefix*postfix = result
        int nextPostfix = 1;
        for(int i = res.length-1; i >= 0; i--){
            res[i] = res[i]*nextPostfix;
            nextPostfix = nums[i]*nextPostfix;
        }
        
        
        return res;
    }
}
        

File: 238-product-of-array-except-self/238-product-of-array-except-self.py
"""
nums: [1, 2, 3, 4]
pre:  [1, 1, 2, 6]
post: [24,12,4,1]
res:  [24,12,8,6]
"""

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        
        #prefix
        res = [1]
        for i in range(len(nums))[1:]:
            res.append(res[i-1] * nums[i-1])
        
        #postfix
        post = 1
        for i in range(len(nums))[::-1]:
            res[i] = res[i] * post
            post = post * nums[i]
        
        return res
        

